quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+        match self {
+            ResolvedCallArgument::Vector { mutable, .. } => *mutable,
+            ResolvedCallArgument::MemoryReference { mutable, .. } => *mutable,
+            ResolvedCallArgument::Immediate { .. } => false,
+        }
+    }
+
+    /// Returns the name of the memory region. In the case of an immediate value,
+    /// this will be `None`.
+    pub(crate) fn name(&self) -> Option<String> {
+        match self {
+            ResolvedCallArgument::Vector {
+                memory_region_name, ..
+            } => Some(memory_region_name.clone()),
+            ResolvedCallArgument::MemoryReference {
+                memory_reference, ..
+            } => Some(memory_reference.name.clone()),
+            ResolvedCallArgument::Immediate { .. } => None,
+        }
+    }
+}
+
+/// A list of arguments for a call instruction. These may be resolved or unresolved.
+/// To resolve a [`Call`] instruction, use [`crate::Program::resolve_call_instructions`].
+#[derive(Clone, Debug, PartialEq, Hash, Eq)]
+pub enum CallArguments {
+    /// The resolved call arguments.
+    Resolved(Vec<ResolvedCallArgument>),
+    /// The unresolved call arguments.
+    Unresolved(Vec<UnresolvedCallArgument>),
+}
+
+impl Quil for CallArguments {
+    fn write(
+        &self,
+        writer: &mut impl std::fmt::Write,
+        fall_back_to_debug: bool,
+    ) -> Result<(), crate::quil::ToQuilError> {
+        match self {
+            CallArguments::Resolved(arguments) => {
+                if !arguments.is_empty() {
+                    write!(writer, " ")?;
+                }
+                for (i, argument) in arguments.iter().enumerate() {
+                    argument.write(writer, fall_back_to_debug)?;
+                    if i < arguments.len() - 1 {
+                        write!(writer, " ")?;
+                    }
+                }
+            }
+            CallArguments::Unresolved(arguments) => {
+                if !arguments.is_empty() {
+                    write!(writer, " ")?;
+                }
+                for (i, argument) in arguments.iter().enumerate() {
+                    argument.write(writer, fall_back_to_debug)?;
+                    if i < arguments.len() - 1 {
+                        write!(writer, " ")?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// An error that can occur when validating a call instruction.
+#[derive(Clone, Debug, PartialEq, thiserror::Error, Eq)]
+pub enum CallValidationError {
+    /// The specified name is not a valid user identifier.
+    #[error(transparent)]
+    Name(#[from] IdentifierValidationError),
+}
+
+/// A call instruction with a name and arguments.
+#[derive(Clone, Debug, PartialEq, Hash, Eq)]
+pub struct Call {
+    /// The name of the call instruction. This must be a valid user identifier.
+    pub name: String,
+    /// The arguments of the call instruction.
+    pub arguments: CallArguments,
```

Comment:
I considered three alternatives here:

1. Just have a `Vec<CallArgument>` where each `CallArgument` has a `resolution: Option<Resolution>` attribute.
2. Do not mutate the `CallArgument`s and instead return a struct that represents the resolution, roughly of the structure ("name", instruction_index) => Vec<ResolvedCallArguments> (we need the `index` to refer to index of the `CALL` instruction within the program).
3. Add a type parameter to the `CALL` instructions and then to the program itself. Resolution would then return a program of a different type (eg `into_call_resolved_program(self) -> Result<Program<ResolveCallArgument>, ProgramError>`).

I landed on the de facto implementation because:

1. Within a given instruction, call arguments are all resolved at the same time. Either all arguments are resolved or none are.
2. There are existing patterns within Quil that mutate the program, such as `resolve_placeholders`.
3. Tracking instruction indices seems fairly unergonomic and brittle.
4. I did not want to add type complexity to the `Program` struct which is pretty easy to use.
5. I wanted to avoid the user resolving the program more than once for efficiency's sake.

I definitely feel like this resolution functionality belongs in quil-rs and not separately in downstream compilers. After going back and forth, I think this is the right implementation, but am open to input here.

Reply:
This is indeed a thorny question!  I agree that alternatives \#2 and (to a lesser extent) \#1 are inferior, but I think there's a lot to be said for \#3.  I've talked to you about this offline, but to expand here:

I think this raises the question of if we've hit the point where we want to separate the parsed AST from the "typechecked"/"resolved" AST.  I think there's a lot of merit to doing so, as it allows for capturing invariants much more cleanly.  But I'm not sure if this PR is the place to do it.

I think I might favor a design where we parameterize all our types by the stage of compilation, passing that down to where we need to make a decision, and default that type parameter to the stage that corresponds to the existing situation.  Something like the following:

```
pub trait Stage {
    type CallArgument;
}

pub enum Parsed;

impl Stage for Parsed {
    type CallArgument = UnresolvedCallArgument;
}

pub enum Resolved;

impl Stage for Resolved {
    type CallArgument = ResolvedCallArgument;
}

pub struct Program<S: Stage = Resolved> {
    // …
    instructions: Instruction<S>,
    // …
}

pub enum Instruction<S: Stage = Resolved> {
    // …
    Call(Call<S>),
    // …
}

pub struct Call<S: Stage = Resolved> {
    pub name: String,
    pub arguments: Vec<S::CallArgument>,
}
```

The upside to this is that we can bundle all the types we need to parameterize by together; the downside is the extra trait.  I think the upside is likely to be worth it, but it's not obvious.

Another limitation of this approach is that it forces the ASTs to be almost identical.  This can be good or bad.  Another approach would be to have

```
pub trait Stage {
    type AST;
}
```

even if `Parsed::AST = Program<α>` and `Resolved::AST = Program<β>` for now.

We can also hide some of the complexity by having the parser return a `Program<Parsed>`, a function `fn resolve(parsed: Program<Parsed>) -> Result<Program<Resolved>, ResolutionError>`, and then exposing at the top level a function that simply combines the two, returning a `Result<Program, …>`, so the user is not confronted with this new API.

That said: this adds extra complexity!  I think that may be worth it, but it's not obvious.  This mutate-and-resolve approach is not a bad one, and it might even be the implementation behind the more complex version I outline above.

Reply:
Just spoke with Kalan on this. We landed somewhere around here:

* `Pragma` doesn't need to be an `enum`. We can add something like `extern_definitions: HashMap<String, ExternDefinition>` to `Program`.
* `struct Call` will only have `arguments: Vec<UnresolvedCallArgument>`. `Call.resolve(...)` returns `Result<Vec<ResolvedCallArgument>, ...>`. This will be a public function for the purposes of translating.

There's a bit of inefficiency here WRT resolving in `get_memory_accesses` (still fallible) and then resolving for translation. I may think a bit more about that, but otherwise, this seems tenable to me.

Reply:
So is the idea here that resolution doesn't need to produce a new `Program`, just store the `extern_definitions`?  And then any processing that needs to consume a resolved `Call` will just call `.resolve` in situ when necessary?  I think this seems fine, if a bit of kicking the can down the road wrt a new AST – but as I said above, this PR is like not the right place for that anyway, so I don't think that's an issue.

Reply:
Yup, you got it. Here I'm just going for consistency with the existing implementation but I'm onboard with your general vision for generating a separate "validated and resolved" AST. We'll keep the conversation going.